File: utils/src/main/java/com/soze/kleddit/utils/sql/DatabaseReset.java
package com.soze.kleddit.utils.sql;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Class used only for system testing,
 * mainly for resetting the database.
 */
public class DatabaseReset {

  private static final String DB_USER = System.getenv("KLEDDIT_DB_USER");
  private static final String DB_PASS = System.getenv("KLEDDIT_DB_PASS");
  private static final String CONNECTION_STRING = "jdbc:postgresql://127.0.0.1:5432/kleddit-test";

  /**
   * This method will load a .sql script located at /sql
   * and execute it. This script will hopefully contain drop / recreate scripts.
   */
  public static void resetDatabase() {

    try {
      List<String> file = Files.lines(Paths.get("/kleddit/sql/create-database.sql"))
        .collect(Collectors.toList());
      String fileContent = file.stream().reduce("", (prev, cur) -> prev += cur + '\n');
      executeSql(fileContent);
    } catch (IOException e) {
      e.printStackTrace();
    }

  }

  private static void executeSql(String sql) {
    try {
      Class.forName("org.postgresql.Driver");
    } catch (ClassNotFoundException e) {
      e.printStackTrace();
    }

    Connection connection = null;

    try {

      connection = DriverManager.getConnection(
        CONNECTION_STRING,
        DB_USER,
        DB_PASS
      );

      PreparedStatement statement = connection.prepareStatement(sql);
      statement.execute();
      System.out.println("CLEARED AND RESTORED THE DB.");

    } catch (SQLException e) {
      e.printStackTrace();
    }

  }


}

File: kleddit-node-server/src/utils/sql/database-reset.ts
import * as pg from 'pg'
import {readFile} from "../file/file-utils";

const DB_USER = process.env['KLEDDIT_DB_USER']
const DB_PASS = process.env['KLEDDIT_DB_PASS']
const HOST = '127.0.0.1'
const DB  = 'kleddit-test' //TODO later load from config

/**
 * Logic of this method is very straightforward, because it's used only
 * for testing.
 * @returns {Promise<void>}
 */
export const resetDatabase = async () => {

    //1. load sql from file
    const sql = await readFile('../sql/create-database.sql')
    //2. connect to db
    const client = new pg.Client({
        user: DB_USER,
        password: DB_PASS,
        host: HOST,
        port: 5432,
        database: DB,
    })

    await client.connect()

    //3. execute
    await client.query(sql)

    console.log('database reset')
    return client.end()
}